### PR TITLE
feat(DEV-12191): improved grid autosize

### PR DESCRIPTION
**Pull request by chebum**
This PR: 
- Reduces amount of layout jumps on page load,
- Preserves column state between page reloads.

### CODE REVIEW

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      );
+    }
+
+    return () => {
+      timeouts.forEach(window.clearTimeout); // Cancel pending timeouts on unmount
+    };
+  }, [gridApiRef, rows, columns, areCounterpartsLoading, serializationKey]);
+
+  // Save columns width when switching between pages
+  useLayoutEffect(() => {
```
Comment (vkuprin) [inline packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:112]: Consider to add early return for trivial cases ` if (!rows || !gridApiRef.current) return; `

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    const grid = gridApiRef.current;
+    const serializedColumnsStr = localStorage.getItem(serializationKey);
```
Comment (vkuprin) [inline packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:114]: I thin in that case we actually should  avoid excessive local storage access. Consider that to wrap in useMemo
` useMemo(() => localStorage.getItem(serializationKey), [serializationKey]);`

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    const timeouts: number[] = [];
+    if (
+      rows && // Do not autosize until rows are loaded
+      rows.length &&
+      grid &&
+      grid.autosizeColumns && // Will be undefined in test environment
+      grid.getRowsCount() && // Make sure the grid was already rendered so that autosizeColumns function will work
+      grid.getCellElement(rows[0].id, grid.getAllColumns()[0].field) && // Make sure the grid was already rendered so that autosizeColumns function will work
+      !areCounterpartsLoading && // Skip autosize until counterparts are loaded
+      !autoSizePerformed.current && // Do not perform autosize twice
```
Comment (vkuprin) [inline packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:55]: Consider setting `autoSizePerformed.current = true` immediately 

```ts

// Before:
if (!autoSizePerformed.current) {
    // Autosize logic
}

// After:
if (!autoSizePerformed.current && !areCounterpartsLoading) {
    autoSizePerformed.current = true;  // Prevent future executions
    // Autosize logic
}
```

 In that case we avoid the potential for multiple executions of autosizing logic if the effect reruns before the current autosize operation completes

**Review by radist2s — APPROVED** [review]
💎

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts (hunk 1 above)
Comment (chebum) [reply packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts]: Checking for rows count makes no sense for deserialization of column's width. I added check for columns number though.

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts (hunk 2 above)
Comment (chebum) [reply packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:118]: We need this value to be sustain page re-loads and component unmounts. useMemo will be cleared on component unmount.

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      !localStorage.getItem(serializationKey) // Do not perform autosize on next page loads - we preserve column state, set by user
+    ) {
+      timeouts.push(
+        window.setTimeout(() => {
+          flushSync(() => {
```
[the remark this answers is not in the document]
Comment (chebum) [reply packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:63]: Yes, I tried. flushSync was added as per DataGrid documentation. I added a link to the corresponding doc on top of the file

File: packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts (hunk 3 above)
Comment (chebum) [reply packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts:57]: autoSizePerformed.current is used for determining if current columns state should be saved. Therefore, this value should be set only after actually performing the autosizing operation

**Pull request merged**